File: lib/providers.ts
import GtfsRealtimeBindings from "gtfs-realtime-bindings";
import type { ServiceAlert, TelemetryRecord } from "./types";
import { log } from "./logging";

const userAgent = "ProjectDispatch/1.0";
const providerTimeoutMs = 15000;
const airports = [
  { code: "PIT", latitude: 40.4915, longitude: -80.2329 },
  { code: "AGC", latitude: 40.3544, longitude: -79.9302 },
  { code: "LBE", latitude: 40.2759, longitude: -79.4048 },
];

async function responseBytes(url: string): Promise<Uint8Array> {
  log.info("Fetching telemetry provider", { provider: "prt", url });
  const response = await fetch(url, { headers: { "User-Agent": userAgent }, cache: "no-store", signal: AbortSignal.timeout(providerTimeoutMs) });
  if (!response.ok) {
    log.warn("Telemetry provider returned an error", { provider: "prt", url, status: response.status });
    throw new Error(`HTTP ${response.status}`);
  }
  return new Uint8Array(await response.arrayBuffer());
}

export async function fetchPrt(url = process.env.PRT_VEHICLE_POSITIONS_URL || "https://truetime.portauthority.org/gtfsrt-bus/vehicles"): Promise<TelemetryRecord[]> {
  let feed;
  try {
    feed = GtfsRealtimeBindings.transit_realtime.FeedMessage.decode(await responseBytes(url));
  } catch (error) {
    throw new Error(`PRT feed was not valid GTFS-RT protobuf: ${error instanceof Error ? error.message : "unknown error"}`);
  }

  const observed = new Date().toISOString();
  return feed.entity.flatMap((entity) => {
    const vehicle = entity.vehicle;
    if (!vehicle) return [];
    const position = vehicle.position;
    return [{
      source: "prt",
      vehicle_id: vehicle.vehicle?.id || entity.id,
      route: vehicle.trip?.routeId || null,
      latitude: position?.latitude ?? null,
      longitude: position?.longitude ?? null,
      speed_mph: position?.speed ? position.speed * 2.23694 : null,
      altitude_ft: null,
      observed_at: observed,
      metadata: {
        current_status: vehicle.currentStatus,
        trip_id: vehicle.trip?.tripId,
        direction_id: vehicle.trip?.directionId,
        current_stop_sequence: vehicle.currentStopSequence,
        current_stop_id: vehicle.stopId,
      },
    }];
  });
}

export async function fetchPrtAlerts(url = process.env.PRT_SERVICE_ALERTS_URL || "https://truetime.portauthority.org/gtfsrt-bus/alerts"): Promise<ServiceAlert[]> {
  log.info("Fetching PRT service alerts", { provider: "prt-alerts", url });
  const response = await fetch(url, { headers: { "User-Agent": userAgent }, cache: "no-store", signal: AbortSignal.timeout(providerTimeoutMs) });
  if (!response.ok) throw new Error(`HTTP ${response.status}`);
  const feed = GtfsRealtimeBindings.transit_realtime.FeedMessage.decode(new Uint8Array(await response.arrayBuffer()));
  return feed.entity.flatMap((entity) => {
    const alert = entity.alert;
    if (!alert) return [];
    const text = (value: unknown) => {
      const translations = (value as { translation?: { text?: string }[] } | undefined)?.translation || [];
      return translations[0]?.text || "";
    };
    const informed = alert.informedEntity || [];
    return [{
      id: entity.id,
      header: text(alert.headerText) || "PRT service alert",
      description: text(alert.descriptionText),
      effect: String(alert.effect || "UNKNOWN"),
      routes: informed.flatMap((item) => item.routeId ? [String(item.routeId)] : []),
      stops: informed.flatMap((item) => item.stopId ? [String(item.stopId)] : []),
      updated_at: new Date().toISOString(),
    }];
  });
}

export async function fetchOpenSky(url = process.env.OPENSKY_STATES_URL || "https://opensky-network.org/api/states/all"): Promise<TelemetryRecord[]> {
  const query = new URLSearchParams({ lamin: "40.2", lamax: "40.7", lomin: "-80.5", lomax: "-79.6" });
  log.info("Fetching telemetry provider", { provider: "opensky", url });
  let response: Response;
  try {
    response = await fetch(`${url}${url.includes("?") ? "&" : "?"}${query}`, { headers: { "User-Agent": userAgent }, cache: "no-store", signal: AbortSignal.timeout(providerTimeoutMs) });
  } catch (error) {
    const cause = error instanceof Error && error.cause instanceof Error ? `; cause: ${error.cause.message}` : "";
    throw new Error(`OpenSky request failed for ${url}: ${error instanceof Error ? error.message : String(error)}${cause}`);
  }
  if (!response.ok) {
    const body = (await response.text()).slice(0, 300);
    log.warn("Telemetry provider returned an error", { provider: "opensky", url, status: response.status });
    throw new Error(`OpenSky HTTP ${response.status}${body ? `: ${body}` : ""}`);
  }
  const payload = await response.json() as { states?: unknown[][] };
  const observed = new Date().toISOString();
  return (payload.states || []).flatMap((state) => {
    if (state.length < 11 || state[5] == null || state[6] == null) return [];
    const latitude = Number(state[6]);
    const longitude = Number(state[5]);
    const airport = airports.find((candidate) => Math.hypot((latitude - candidate.latitude) * 69, (longitude - candidate.longitude) * 53) <= 25);
    if (!airport) return [];
    return [{
      source: "opensky",
      vehicle_id: String(state[1] || state[0]),
      route: null,
      latitude,
      longitude,
      speed_mph: state[9] == null ? null : Number(state[9]) * 2.23694,
      altitude_ft: state[7] == null ? null : Number(state[7]) * 3.28084,
      observed_at: observed,
      metadata: { callsign: String(state[1] || "").trim(), on_ground: state[8], airport: airport.code },
    }];
  });
}

export async function fetchAmtrak(): Promise<TelemetryRecord[]> {
  const url = process.env.AMTRAK_TELEMETRY_URL;
  if (!url) {
    log.warn("Amtrak telemetry is not configured; no train positions will be shown");
    return [];
  }
  log.info("Fetching telemetry provider", { provider: "amtrak", url });
  const response = await fetch(url, { headers: { "User-Agent": userAgent }, cache: "no-store", signal: AbortSignal.timeout(providerTimeoutMs) });
  if (!response.ok) {
    log.warn("Telemetry provider returned an error", { provider: "amtrak", url, status: response.status });
    throw new Error(`HTTP ${response.status}`);
  }
  const payload = await response.json() as unknown;
  const items = Array.isArray(payload) ? payload : ((payload as { trains?: unknown[] }).trains || []);
  return items.flatMap((value) => {
    const item = value as Record<string, unknown>;
    const number = String(item.trainNumber ?? item.number ?? "");
    if (!["42", "43"].includes(number)) return [];
    return [{
      source: "amtrak",
      vehicle_id: number,
      route: "Pennsylvanian",
      latitude: typeof item.latitude === "number" ? item.latitude : null,
      longitude: typeof item.longitude === "number" ? item.longitude : null,
      speed_mph: typeof item.speed === "number" ? item.speed : null,
      altitude_ft: null,
      observed_at: new Date().toISOString(),
      metadata: item,
    }];
  });
}

File: backend/providers.py
from __future__ import annotations

import json
import logging
from datetime import datetime, timezone
from typing import Any
from urllib.parse import urlencode
from urllib.request import Request, urlopen

from .models import TelemetryRecord

logger = logging.getLogger(__name__)


def fetch_json(url: str, *, timeout: float = 10) -> Any:
    request = Request(url, headers={"User-Agent": "ProjectDispatch/1.0"})
    with urlopen(request, timeout=timeout) as response:
        return json.loads(response.read())


class PRTProvider:
    """Reads PRT GTFS-RT vehicle positions when gtfs-realtime-bindings is installed."""

    default_url = "https://truetime.portauthority.org/gtfsrt/vehiclePositions"

    def __init__(self, url: str | None = None) -> None:
        self.url = url or self.default_url

    def fetch(self) -> list[TelemetryRecord]:
        try:
            from google.transit import gtfs_realtime_pb2
            from google.protobuf.message import DecodeError
        except ImportError:
            logger.warning("PRT polling skipped: install gtfs-realtime-bindings to parse protobuf")
            return []

        request = Request(self.url, headers={"User-Agent": "ProjectDispatch/1.0"})
        with urlopen(request, timeout=10) as response:
            feed = gtfs_realtime_pb2.FeedMessage()
            try:
                feed.ParseFromString(response.read())
            except DecodeError as exc:
                raise ValueError("PRT response was not a valid GTFS-RT protobuf feed") from exc

        observed = datetime.now(timezone.utc)
        records: list[TelemetryRecord] = []
        for entity in feed.entity:
            if not entity.HasField("vehicle"):
                continue
            vehicle = entity.vehicle
            position = vehicle.position if vehicle.HasField("position") else None
            records.append(
                TelemetryRecord(
                    source="prt",
                    vehicle_id=vehicle.vehicle.id or entity.id,
                    route=vehicle.trip.route_id or None,
                    latitude=position.latitude if position else None,
                    longitude=position.longitude if position else None,
                    speed_mph=(position.speed * 2.23694) if position and position.speed else None,
                    altitude_ft=None,
                    observed_at=observed,
                    metadata={"current_status": vehicle.current_status},
                )
            )
        return records


class OpenSkyProvider:
    default_url = "https://opensky-network.org/api/states/all"
    # A small PIT bounding box keeps the response useful and inexpensive.
    bounds = {"lamin": 40.2, "lamax": 40.7, "lomin": -80.5, "lomax": -79.6}

    def __init__(self, url: str | None = None) -> None:
        self.url = url or self.default_url

    def fetch(self) -> list[TelemetryRecord]:
        url = f"{self.url}?{urlencode(self.bounds)}"
        payload = fetch_json(url)
        observed = datetime.now(timezone.utc)
        records: list[TelemetryRecord] = []
        for state in payload.get("states") or []:
            if len(state) < 11 or state[5] is None or state[6] is None:
                continue
            records.append(
                TelemetryRecord(
                    source="opensky",
                    vehicle_id=(state[1] or state[0]).strip(),
                    route=None,
                    latitude=state[6],
                    longitude=state[5],
                    speed_mph=(state[9] * 2.23694) if state[9] is not None else None,
                    altitude_ft=state[7] * 3.28084 if state[7] is not None else None,
                    observed_at=observed,
                    metadata={"callsign": (state[1] or "").strip(), "on_ground": state[8]},
                )
            )
        return records


class AmtrakProvider:
    """Adapter for the unofficial Track Your Train JSON endpoint.

    The endpoint is intentionally configurable because Amtrak can change its map
    implementation without notice. An empty URL disables this provider.
    """

    def __init__(self, url: str | None = None) -> None:
        self.url = url

    def fetch(self) -> list[TelemetryRecord]:
        if not self.url:
            return []
        payload = fetch_json(self.url)
        items = payload if isinstance(payload, list) else payload.get("trains", [])
        records: list[TelemetryRecord] = []
        for item in items:
            if str(item.get("trainNumber", item.get("number", ""))) not in {"42", "43"}:
                continue
            records.append(
                TelemetryRecord(
                    source="amtrak",
                    vehicle_id=str(item.get("trainNumber", item.get("number"))),
                    route="Pennsylvanian",
                    latitude=item.get("latitude"),
                    longitude=item.get("longitude"),
                    speed_mph=item.get("speed"),
                    altitude_ft=None,
                    observed_at=datetime.now(timezone.utc),
                    metadata=item,
                )
            )
        return records
